# Контрольная работа
Сафонов Глеб

In [1]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
import statsmodels.api as sm

## Задача 1

In [2]:
df_sleep = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/sleep75.csv')
# специфицируем модель через формулу
sleep_eq1 = smf.ols(formula='sleep~age+smsa', data=df_sleep).fit()
# Коэффициенты модели с округлением
sleep_eq1.params.round(2)

Intercept    3151.38
age             3.61
smsa          -62.83
dtype: float64

## Задача 2

Вычислим $R^2$ и $R^2_{adj}$

In [6]:
df_wage = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/wage2.csv')
# специфицируем модель через формулу
wage_eq1 = smf.ols(formula='np.log(wage)~age+south+urban+married', data=df_wage).fit()
R2 = wage_eq1.rsquared
R2_adj = wage_eq1.rsquared_adj
print("R2 =", R2.round(3))
print("R2_adj =", R2_adj.round(3))

R2 = 0.118
R2_adj = 0.114


In [8]:
# Коэффициенты модели с округлением
wage_eq1.params.round(2)

Intercept    5.89
age          0.02
south       -0.15
urban        0.18
married      0.20
dtype: float64

## Задача 3

Вычислим TSS, ESS, RSS

In [9]:
df_labour = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/Labour.csv')
# Подгонка модели
output_eq1 = smf.ols(formula='np.log(output)~np.log(capital)+np.log(wage)', data=df_labour).fit()
output_eq1.params.round(2)

Intercept         -1.94
np.log(capital)    0.51
np.log(wage)       0.88
dtype: float64

In [10]:
TSS = output_eq1.centered_tss
ESS = output_eq1.ess
RSS = output_eq1.ssr
print("TSS =", TSS.round(2))
print("ESS =", ESS.round(2))
print("RSS =", RSS.round(2))

TSS = 797.67
ESS = 487.23
RSS = 310.43


## Задача 4

In [11]:
# специфицируем модель через формулу
sleep_eq2 = smf.ols(formula='sleep~totwrk+age+male+smsa', data=df_sleep).fit()
# Коэффициенты модели с округлением
sleep_eq2.params.round(2)

Intercept    3494.22
totwrk         -0.17
age             2.81
male           86.91
smsa          -75.29
dtype: float64

Для наблюдения с номером 3 найдем фактическое значение зависимой переменной, предсказанное значение и остаток

In [16]:
# Фактичечкое
fact = df_sleep["sleep"][3].round(3)
# Предсказанное
pred = sleep_eq2.fittedvalues[3].round(3)
# Остаток
resid = sleep_eq2.resid[3].round(3)
print("Фактичечкое =", fact)
print("Предсказанное =", pred)
print("Остаток =", resid)

Фактичечкое = 3083
Предсказанное = 2943.58
Остаток = 139.42


## Задача 5

Для набора данных 'Labour' рассмотрим линейную регрессию **log(output) на log(capital), labour, wage**

Оценим регрессию на данных

In [17]:
# Подгонка модели
labour_eq = smf.ols(formula='np.log(output)~np.log(capital)+labour+wage', data=df_labour).fit()
labour_eq.params.round(2)

Intercept          0.58
np.log(capital)    0.44
labour             0.00
wage               0.02
dtype: float64

Для некоторых наблюдений найдем фактическое значение зависимой переменной, предсказанное значение и остатки.

In [18]:
ind = np.array([6, 37, 68, 139, 365, 479])-1
df = pd.DataFrame({'i': ind + 1, 'Факт': np.log(df_labour["output"][ind]), 'Fitted': labour_eq.predict()[ind], 'Остаток': np.log(df_labour["output"][ind]) - labour_eq.predict()[ind]})
df

i      Факт    Fitted   Остаток
5      6  6.174079  7.298163 -1.124085
36    37  1.399213  2.000322 -0.601109
67    68  1.586001  2.168943 -0.582942
138  139  0.337593  1.205367 -0.867774
364  365  1.727330  1.432167  0.295163
478  479  1.547772  1.934992 -0.387220

Вычислим TSS, ESS, RSS

In [21]:
TSS = labour_eq.centered_tss
ESS = labour_eq.ess
RSS = labour_eq.ssr
print("TSS =", TSS.round(3))
print("ESS =", ESS.round(3))
print("RSS =", RSS.round(3))

TSS = 797.667
ESS = 522.54
RSS = 275.128


Вычислим $R^2$ и $R^2_{adj}$

In [22]:
R2 = labour_eq.rsquared
R2_adj = labour_eq.rsquared_adj
print("R2 =", R2.round(3))
print("R2_adj =", R2_adj.round(3))

R2 = 0.655
R2_adj = 0.653


Спецификация: $\log(output)=0.58+0.44*\log(capital)+0*labour+0.02*wage+u$

### Интерпретация модели
1) При увеличении капитала на 1%, выпуск увеличивается на 0.44%
2) При увеличении числа рабочих на 1, выпуск не меняется